contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -2088,6 +2089,292 @@ numeric_radians(PG_FUNCTION_ARGS)
+					 errmsg("invalid UTF8 byte sequence starting with 0x%02x",
+							in[0])));
+		code = ((in[0] & 0x07) << 18) | ((in[1] & 0x3F) << 12) |
+			((in[2] & 0x3F) << 6) | (in[3] & 0x3F);
+		consumed = 4;
+	}
+	else
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_DATA_EXCEPTION),
+				 errmsg("invalid UTF8 byte sequence starting with 0x%02x",
+						in[0])));
+	}
+
+	if (code > 0x10FFFF || (code >= 0xD800 && code < 0xE000))
+		ereport(ERROR,
+				(errcode(ERRCODE_DATA_EXCEPTION),
+				 errmsg("invalid UTF8 code point 0x%x", code)));
+
+	if (consumed_p)
+		*consumed_p = consumed;
+
+	return code;
+}
+
+Datum
+parsename(PG_FUNCTION_ARGS)
+{
+    text *object_name = PG_GETARG_TEXT_PP(0);
+    int object_piece = PG_GETARG_INT32(1);
+    char *object_name_str = text_to_cstring(object_name);
+    int len = strlen(object_name_str);
+    int state = 0;
+    int consumed;
+    int32_t code;
+    int total_chars = 0;
+    int total_length = 0;
+    char c;
+    char *start_positions[4] = {NULL};
+    char *end_positions[4] = {NULL};
+    int initial_state[4] = {0};
+    int current_part = 0;
+    text *result;
+    start_positions[current_part] = object_name_str;
+	
+    // object_piece should only have maximum of 4 parts.
+    if (object_piece < 1 || object_piece > 4)
+    {
+        PG_RETURN_NULL();
+    }
+
+    for (int i = 0; i < len;)
+    {
+        code = GetUTF8CodePoint((const unsigned char *)&object_name_str[i], len - i, &consumed);
+        c = object_name_str[i];
+        if (total_chars > 128 || total_length > 256)
+        {
+            PG_RETURN_NULL();
+        }
+
+        if (state == 0)
+        {
+            if (c == '"')
+            {
+                state = 1;
+                // save the initial state so that we can escape the correct characters at the end.
+                if (initial_state[current_part] == 0)
+                {
+                    initial_state[current_part] = 1;
+                }
+
+                start_positions[current_part] = &object_name_str[i + 1];
+                i += consumed;
+                continue;
+            }
+            else if(c == ']')
+            {
+                pfree(object_name_str);
+                PG_RETURN_NULL();
+            }
+            else if (c == '[')
+            {
+                state = 2;
+                if (initial_state[current_part] == 0)
+                {
+                    initial_state[current_part] = 2;
+                }
+
+                start_positions[current_part] = &object_name_str[i + 1];
+                i += consumed;
+                continue;
+            }
+            else if (c == '.')
+            {
+                // do not update the value of end_positions[current_part] if there is already a value in end_postions[current_part] & previous character is " or ].
+                if ( !((end_positions[current_part] != NULL) && (object_name_str[i - 1] == '"')) && !((end_positions[current_part] != NULL) && (object_name_str[i - 1] == ']')) )
```

Comment:
I feel this condition is very confusing. Can the program reach line 2248? Under which input?